shell/server/core.js

```diff
@@ -158,17 +187,46 @@ class NotificationHandle {
 }
 
 class PersistentUiViewImpl {
-  constructor(persistentMethods) {
+  constructor(persistentMethods, grain) {
+    this._grain = grain;
     _.extend(this, persistentMethods);
   }
 
+  getViewInfo() {
+    const viewInfo = this._grain.cachedViewInfo;
+    return inMeteor(() => {
+      const pkg = Packages.findOne({ _id: this._grain.packageId });
+      const manifest = pkg.manifest || {};
+      if (!viewInfo.metadata) {
+        viewInfo.metadata = {};
+      }
+
+      if (!viewInfo.metadata.appTitle) {
+        viewInfo.metadata.appTitle = manifest.appTitle || {};
+      }
+
+      if (!viewInfo.metadata.icon && !viewInfo.metadata.appId) {
+        const grainIcon = ((manifest.metadata || {}).icons || {}).grain;
+        if (grainIcon) {
+          viewInfo.metadata.icon = grainIcon;
+        } else {
+          viewInfo.metadata.appId = pkg.appId;
+        }
+      }
+
+      return viewInfo;
+    });
+  }
+
   // All other UiView methods are currently unimplemented, which, while not strictly correct,
   // results in the same overall behavior, since users can't call restore() on a PersistentUiView,
   // and grains can't call methods on UiViews because they lack the "is human" pseudopermission.
 }
 
-const makePersistentUiView = function (persistentMethods) {
-  return new Capnp.Capability(new PersistentUiViewImpl(persistentMethods), PersistentUiView);
+const makePersistentUiView = function (persistentMethods, grainId) {
+  check(grainId, String);
+  const grain = Grains.findOne({ _id: grainId });
```

Comment:
Good point. I think makes sense to also do the query here so we can throw an error early if the grain has been deleted.